src/mmw/apps/modeling/mapshed/tasks.py

```diff
@@ -457,30 +458,58 @@ def nlcd_kfactor(result):
-    definitions = [
-        ('nlcd_streams', {'polygon': [geom.geojson],
-                          'vector': streams(geom)}, nlcd_streams),
-        ('nlcd_soils', {'polygon': [geom.geojson]}, nlcd_soils),
-        ('gwn', {'polygon': [geom.geojson]}, gwn),
-        ('avg_awc', {'polygon': [geom.geojson]}, avg_awc),
-        ('nlcd_slope', {'polygon': [geom.geojson]}, nlcd_slope),
-        ('slope', {'polygon': [geom.geojson]}, slope),
-        ('nlcd_kfactor', {'polygon': [geom.geojson]}, nlcd_kfactor)
-    ]
-
-    tasks = zip(definitions, [choose_worker()
-                              for _ in xrange(len(definitions))])
-
-    return [(mapshed_start.s(opname, data).set(exchange=exchange,
-                                               routing_key=geop_worker) |
-             mapshed_finish.s().set(exchange=exchange,
-                                    routing_key=geop_worker) |
-             callback.s().set(link_error=errback,
-                              exchange=exchange,
-                              routing_key=choose_worker()))
-            for ((opname, data, callback), geop_worker) in tasks]
+def geop_tasks():
+    return geop_task_defs().keys()
+
+
+def geop_task_defs():
+    return {
+        'nlcd_streams': lambda geom: (
+            'nlcd_streams',
+            {'polygon': [geom.geojson], 'vector': streams(geom)},
+            nlcd_streams),
+        'nlcd_soils': lambda geom: (
+            'nlcd_soils',
+            {'polygon': [geom.geojson]},
+            nlcd_soils),
+        'gwn': lambda geom: ('gwn', {'polygon': [geom.geojson]}, gwn),
+        'avg_awc': lambda geom: (
+            'avg_awc',
+            {'polygon': [geom.geojson]},
+            avg_awc),
+        'nlcd_slope': lambda geom: (
+            'nlcd_slope',
+            {'polygon': [geom.geojson]},
+            nlcd_slope),
+        'slope': lambda geom: (
+            'slope',
+            {'polygon': [geom.geojson]},
+            slope),
+        'nlcd_kfactor': lambda geom: (
+            'nlcd_kfactor',
+            {'polygon': [geom.geojson]},
+            nlcd_kfactor)
+    }
+
+
+@shared_task(bind=True, default_retry_delay=1, max_retries=42)
+def geop_task(self, taskName, geom, exchange, errback, choose_worker):
+    try:
+        (opname, data, callback) = geop_task_defs()[taskName](geom)
+        geop_worker = choose_worker()
+        return (mapshed_start.s(opname, data).set(exchange=exchange,
+                                                  routing_key=geop_worker) |
+                mapshed_finish.s().set(exchange=exchange,
+                                       routing_key=geop_worker) |
+                callback.s().set(link_error=errback,
+                                 exchange=exchange,
+                                 routing_key=geop_worker))
+    except Retry as r:
+        raise r
+    except Exception as x:
+        return {
+            'error': x.message
+        }
```

Comment:
Regarding swallowing exceptions: this was done in the first two tasks in the three-task chains here:

![mmw celery workflow](https://cloud.githubusercontent.com/assets/1430060/19315847/50266f98-906d-11e6-90b2-d9d4dffcbb18.png)

Because if there was an exception in the inner chain on any task except the last task, the chord would never unlock and Celery would be stuck in a loop forever.

Here's an excerpt from my upcoming blog post about the matter:

> #### Error Handling in Nested Chains
> Even though the MapShed workflow shown above is complex with multiple specialized tasks, the error handling is very general. Regardless of the task that raises an exception or encounters an error, we simply want to log it to the database and return to the client. We assumed that given its general applicability, we could specify the error handler once on the parent chain and have it be catch errors raised anywhere in the constituent tasks. Unfortunately this is not true, and we must attach the error handler separately to every task in every subchain. We do this by [defining the error handling task once](https://github.com/WikiWatershed/model-my-watershed/blob/3c4f7e2123481d9909d18d1dad9699ed3c24aebe/src/mmw/apps/modeling/views.py#L257-L258), and [passing it as a parameter](https://github.com/WikiWatershed/model-my-watershed/blob/3c4f7e2123481d9909d18d1dad9699ed3c24aebe/src/mmw/apps/modeling/views.py#L263) to the function that [creates all geoprocessing task chains](https://github.com/WikiWatershed/model-my-watershed/blob/3c4f7e2123481d9909d18d1dad9699ed3c24aebe/src/mmw/apps/modeling/mapshed/tasks.py#L488).